File: common/src/main/java/com/bondhub/common/dto/ApiResponse.java
package com.bondhub.common.dto;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;

import java.util.Map;

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
public record ApiResponse<T>(int code, String message, T data, Map<String, String> errors) {
    public static <T> ApiResponse<T> success(T data) {
        return new ApiResponse<>(1000, "Successful", data, null);
    }

    public static <T> ApiResponse<T> successWithoutData() {
        return new ApiResponse<>(1000, "Successful", null, null);
    }

    public static <T> ApiResponse<T> error(int code, String message, Map<String, String> errors) {
        return new ApiResponse<>(code, message, null, errors);
    }
}

File: post-recommendation-service/app/common/api_response.py
from typing import Generic, TypeVar

from pydantic import BaseModel

T = TypeVar("T")


class ApiResponse(BaseModel, Generic[T]):
    code: int
    message: str
    data: T | None = None
    errors: dict[str, str] | None = None

    @staticmethod
    def success(data: T) -> "ApiResponse[T]":
        return ApiResponse(code=1000, message="Successful", data=data, errors=None)

    @staticmethod
    def success_without_data() -> "ApiResponse[None]":
        return ApiResponse(code=1000, message="Successful", data=None, errors=None)

    @staticmethod
    def error(
        code: int,
        message: str,
        errors: dict[str, str] | None = None,
    ) -> "ApiResponse[None]":
        return ApiResponse(code=code, message=message, data=None, errors=errors)
